dali/util/half.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_UTIL_HALF_H_
+#define DALI_UTIL_HALF_H_
+
+// Macros for writing RunImpl templated code for derived operators
+template <typename Out>
```

Comment:
As described in an e-mail, this can be specialized (depending on CPU/CUDA platform) for half_float::half and __half type. We should probably also expose those types as float16.

Reply:
"float16" is already taken by __half. I exposed them as "_float16"

Reply:
I exposed half_float::half , __half and uint16_t ( which is used in F16C intrinsics as _float16.
I could not find the way to expose all three of them as "float16" due to the compilation problem of *.cu files.